chain/tendermint/src/chain.rs

```diff
@@ -0,0 +1,295 @@
+}
+
+pub struct TriggersAdapter {}
+
+#[async_trait]
+impl TriggersAdapterTrait<Chain> for TriggersAdapter {
+    async fn scan_triggers(
+        &self,
+        _from: BlockNumber,
+        _to: BlockNumber,
+        _filter: &TriggerFilter,
+    ) -> Result<Vec<BlockWithTriggers<Chain>>, Error> {
+        panic!("Should never be called since not used by FirehoseBlockStream")
+    }
+
+    async fn triggers_in_block(
+        &self,
+        _logger: &Logger,
+        block: codec::EventList,
+        _filter: &TriggerFilter,
+    ) -> Result<BlockWithTriggers<Chain>, Error> {
+        let shared_block = Arc::new(block.clone());
+
+        let mut triggers: Vec<_> = shared_block
+            .events()
+            .into_iter()
+            .map(|event| {
+                TendermintTrigger::Event(Arc::new(codec::EventData {
+                    event: Some(event),
+                    block: Some(block.block().clone()),
+                }))
+            })
+            .collect();
+
+        triggers.push(TendermintTrigger::Block(shared_block.cheap_clone()));
+
+        Ok(BlockWithTriggers::new(block, triggers))
+    }
+
+    async fn is_on_main_chain(&self, _ptr: BlockPtr) -> Result<bool, Error> {
+        panic!("Should never be called since not used by FirehoseBlockStream")
+    }
+
+    fn ancestor_block(
+        &self,
+        _ptr: BlockPtr,
+        _offset: BlockNumber,
+    ) -> Result<Option<codec::EventList>, Error> {
+        panic!("Should never be called since not used by FirehoseBlockStream")
+    }
+
+    /// Panics if `block` is genesis.
+    /// But that's ok since this is only called when reverting `block`.
+    async fn parent_ptr(&self, block: &BlockPtr) -> Result<Option<BlockPtr>, Error> {
+        Ok(Some(BlockPtr {
+            hash: BlockHash::from(vec![0xff; 32]),
+            number: block.number.saturating_sub(1),
+        }))
+    }
+}
+
+pub struct FirehoseMapper {}
+
+#[async_trait]
+impl FirehoseMapperTrait<Chain> for FirehoseMapper {
+    async fn to_block_stream_event(
+        &self,
+        logger: &Logger,
+        response: &firehose::Response,
+        adapter: &TriggersAdapter,
+        filter: &TriggerFilter,
+    ) -> Result<BlockStreamEvent<Chain>, FirehoseError> {
+        let step = ForkStep::from_i32(response.step).unwrap_or_else(|| {
+            panic!(
+                "unknown step i32 value {}, maybe you forgot update & re-regenerate the protobuf definitions?",
+                response.step
+            )
+        });
+
+        let any_block = response
+            .block
+            .as_ref()
+            .expect("block payload information should always be present");
+
+        // Right now, this is done in all cases but in reality, with how the BlockStreamEvent::Revert
+        // is defined right now, only block hash and block number is necessary. However, this information
+        // is not part of the actual bstream::BlockResponseV2 payload. As such, we need to decode the full
```

Comment:
@maoueh could this boilerplate be extracted somewhere else for every Firehose chain.